Added emoji picker in mobile app keyboard topbar

harshal2030 opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A brief description of what this pull request does.
-->
This PR adds emoji picker in the quick actions and merge library option with camera using bottom sheet
#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket or fixes a reported issue, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-mobile/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
Fixes https://github.com/mattermost/mattermost/issues/24323
Fixes https://github.com/mattermost/mattermost/issues/24460
#### Checklist
<!--
Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.
-->
- [x] Has UI changes
- [x] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: <!-- Device name(s), OS version(s) -->
Device Name: Samsung Galaxy S20 FE
OS: Android 13
#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs/Videos (for both iOS and Android if possible).
-->

https://github.com/mattermost/mattermost-mobile/assets/43959510/e58879e5-5b0f-4f0f-9fc7-5cde03706490

#### Release Note
<!--
Add a release note for each of the following conditions:

* New features and improvements, including behavioural changes, UI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Example:

```release-note
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->

```release-note
Adds emoji picker in the quick action and image options are merged in one bottom sheet.
```

mattermost-build commented:
Hello @harshal2030,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

harshal2030 commented:
I believe that it needs change with localization files if needed exact text from figma for image bottom sheet. I need input from your side on this, if similar text from localization can be used or needs a change.

thanks

larkox commented on app/components/post_draft/quick_actions/input_quick_action/index.tsx:
```diff
@@ -35,27 +43,65 @@ export default function InputQuickAction({
-    const actionTestID = disabled ?
-        `${testID}.disabled` :
-        testID;
+    const handleEmojiClick = useCallback((emoji: string) => {
+        let emojiDraft: string;
+
+        const emojiData = getEmojiByName(emoji, customEmojis);
+
+        if (emojiData?.image && emojiData.category !== 'custom') {
+            const codeArray: string[] = emojiData.image.split('-');
+            const code = codeArray.reduce((acc, c) => {
+                return acc + String.fromCodePoint(parseInt(c, 16));
+            }, '');
+            emojiDraft = code;
+        } else {
+            emojiDraft = `:${emoji}: `;
+        }
```
A great part of this logic probably comes from `components/autocomplete/emoji_suggestion/emoji_suggestion.tsx`. Any chance we can extract it to a common function in the utils folder?

harshal2030 replied:
yeah, I feel too that it can be refactored, will do it

larkox reviewed (changes requested):
LGTM. The only required change is the one about not needing the `customEmojis`. The rest I leave it to you to consider whether it make sense to address it here.

larkox commented on app/components/post_draft/quick_actions/image_quick_action/index.tsx:
```diff
@@ -4,15 +4,22 @@
+import {bottomSheetSnapPoint} from '@utils/helpers';
 import {changeOpacity} from '@utils/theme';
 
+import CameraType, {type ImageOptions} from './image_type';
```
The file has been renamed to image type, but you are still importing it here as camera type.

larkox reviewed (commented):
Another minor thing.

larkox commented:
@harshal2030 Regarding the strings, you need to run `npm run i18n-extract`. That will add the new string to the `en.json`. Then there is the question about the changes on the strings.

@abhijit-singh The figma design had a different text than what we have nowadays in the app:
`Capture Photo` vs `Take a photo`
`Record Video` vs `Record a video`
Is this something we want to update?

If so, @harshal2030 you will need to update those two strings manually on `en.json`. The rest of the languages will be taken care of in weblate.

harshal2030 commented:
Ok cook

harshal2030 commented on app/components/post_draft/quick_actions/input_quick_action/index.tsx:
```diff
@@ -35,27 +43,65 @@ export default function InputQuickAction({
+            emojiDraft = `:${emoji}: `;
+        }
+
+        updateValue((v) => `${v}${emojiDraft} `);
```
ok, if I would like to work on that, would be really helpful if you can automatically assign me when issue created

abhijit-singh commented:
@harshal2030 & @larkox We don't need to update the strings right now. What we have in the app currently works 👍

harshal2030 commented:
@larkox I've updated code based on your comments also I found some bugs along the way such wrong cursor position on autocomplete for both @ mention and emoji. So I fixed those too in this PR. I'm attaching the screenshot before the fix to have better context.
![emoji](https://github.com/mattermost/mattermost-mobile/assets/43959510/71200646-d849-4784-a58a-ab3d8970f44b)
![at_mention](https://github.com/mattermost/mattermost-mobile/assets/43959510/f0f69212-5144-4172-9252-82a85901dbd9)

larkox commented on app/components/post_draft/quick_actions/image_quick_action/image_type.tsx:
```diff
@@ -65,20 +82,26 @@ const CameraType = ({onPress}: Props) => {
                 onPress={onPhoto}
                 testID='camera_type.photo'
-                text={intl.formatMessage({id: 'camera_type.photo.option', defaultMessage: 'Capture Photo'})}
+                text={intl.formatMessage({id: 'camera_type.photo.option', defaultMessage: 'Capture photo'})}
```
You have an unexpected change here and the next component also.

larkox commented on app/components/post_draft/quick_actions/input_quick_action/index.tsx:
```diff
@@ -34,28 +41,78 @@ export default function InputQuickAction({
+            let emojiDraft: string;
+            const {emojiCode, emojiData} = getEmojiCodeAndData(emoji, []);
+
+            if (emojiData?.image && emojiData.category !== 'custom') {
```
We already have this if inside `getEmojiCodeAndData` so I wonder if there is any "readable and nice" way to get rid of it.

I understand we still need to know if it is a unicode emoji or just text to add or not the `:`, and that we don't add the `:` consistently in the two places this function is called, so... Happy to leave it as it is, but if you come up with a cleaner way, I am all ears :smile:

harshal2030 replied:
ok, I'll try

larkox commented on app/components/post_draft/post_input/post_input.tsx:
```diff
@@ -319,6 +319,7 @@ export default function PostInput({
             onChangeText={handleTextChange}
             onFocus={onFocus}
             onPaste={onPaste}
+            selection={{start: cursorPosition, end: cursorPosition}}
```
Not sure what this change is for.

harshal2030 replied:
Its for changing cursor position

harshal2030 replied:
> @larkox I've updated code based on your comments also I found some bugs along the way such wrong cursor position on autocomplete for both @ mention and emoji. So I fixed those too in this PR. I'm attaching the screenshot before the fix to have better context. ![emoji](https://user-images.githubusercontent.com/43959510/265125009-71200646-d849-4784-a58a-ab3d8970f44b.jpg) ![at_mention](https://user-images.githubusercontent.com/43959510/265125016-f0f69212-5144-4172-9252-82a85901dbd9.jpg)

If we select an option in autocomplete when the cursor is between two text, the position of cursor not correct.

larkox replied:
I wonder if this will mess with selections in general, since (if I am reading right) if I select some text, this will overwrite that selection, and leave it unselected. Am I right?

harshal2030 replied:
Yes right, I'll fix it

larkox reviewed (approved):
A few minor things, but LGTM. Great work!

harshal2030 commented:
@larkox fixed your comments, fixed selection prop too, still looking for better refactor

larkox commented on app/components/post_draft/post_input/post_input.tsx:
```diff
@@ -319,6 +321,7 @@ export default function PostInput({
             onChangeText={handleTextChange}
             onFocus={onFocus}
             onPaste={onPaste}
+            selection={canUpdateCursorPosition ? {start: cursorPosition, end: cursorPosition} : undefined}
```
I don't completely get the idea here. Can you explain it?

I see that you set the `canUpdateCursorPosition` state based on who is updating the cursor position, but it would feel that the moment one of the systems sets this to true, we stop being able to select some part of the input.

harshal2030 replied:
So, the need of updating cursor is when the user uses autocomplete or pick emoji. Whenever user, does the same, `canUpdateCursorPosition` sets to true but then after when user selects or type something, it changes to false and we no longer control selection. This way the selection doesn't gets over written

larkox replied:
@yasserfaraazkhan Please give some love to this part of the functionality when you get to QA this.

enahum commented:
/update-branch

harshal2030 commented:
@larkox I've added one more issue to this PR, please review and let me know if anything needs change

mattermost-build commented:
Building app in separate branch.

enahum commented on app/utils/emoji/helpers.ts:
```diff
@@ -370,3 +369,24 @@ export const searchEmojis = (fuse: Fuse<string>, searchTerm: string) => {
     return [];
 };
+
+export const getEmojiCode = (emoji: string, customEmojis: CustomEmojiModel[] = [], ios = false) => {
```
the last argument can be removed and use the `Platform` module to identify if is iOS

enahum commented on app/utils/emoji/helpers.ts:
```diff
@@ -370,3 +369,24 @@ export const searchEmojis = (fuse: Fuse<string>, searchTerm: string) => {
+
+    // We are going to set a double : on iOS to prevent the auto correct from taking over and replacing it
+    // with the wrong value, this is a hack but I could not found another way to solve it
+    const prefix = ios ? '::' : ':';
```
```suggestion
    const prefix = Platform.OS === 'ios' ? '::' : ':';
```

enahum commented on app/components/autocomplete/emoji_suggestion/emoji_suggestion.tsx:
```diff
@@ -127,27 +127,17 @@ const EmojiSuggestion = ({
-        if (Platform.OS === 'ios') {
-            prefix = '::';
-        }
+        const isIOS = Platform.OS === 'ios';
```
not needed, see the comment on `getEmojiCode` utility function

enahum commented on app/components/post_draft/quick_actions/image_quick_action/image_type.tsx:
```diff
@@ -28,28 +28,45 @@ const getStyle = makeStyleSheetFromTheme((theme: Theme) => ({
+    const onLibrary = async () => {
+        const options: ImageOptions = {
+            mediaType: 'photo',
+            saveToPhotos: true,
```
this is not needed

harshal2030 commented:
@enahum fixed your comments, let me know if anything needs change

enahum commented on app/components/post_draft/quick_actions/image_quick_action/image_type.tsx:
```diff
@@ -65,6 +80,12 @@ const CameraType = ({onPress}: Props) => {
+            <SlideUpPanelItem
+                icon='image-outline'
+                onPress={onLibrary}
+                testID=''
```
Why empty?

harshal2030 replied:
added

harshal2030 commented:
Hi, there's been no update on this PR for a long time, is there anything I need to change?

enahum reviewed (commented):
Is this PR missing the check for the config to determine if the emoji picker is enabled or not? I couldn't find it anywhere

harshal2030 commented:
ok, got it

abhijit-singh reviewed (changes requested):
Thanks @harshal2030! Awesome work! The basic functionality looks good but I can see a few weird issues here and there. Will try to explain them here and also attaching a screencap:

* After adding an emoji from the new emoji picker for the first time once you open a particular  channel, the input cursor lands before the emoji. Expectation - the cursor should be placed after the emoji. 
* Also, if you add more emojis after that, the placement is not consistent. They should ideally be placed one after the other. (Notice where each successive emoji gets added in the screencap attached)
* After adding an emoji, if you try to type something, the first character gets deleted automatically. 

https://github.com/mattermost/mattermost-mobile/assets/14369607/a0126d9f-8e27-4eb1-a46d-942a8f905b66


https://github.com/mattermost/mattermost-mobile/assets/14369607/e2f88a3d-cbd4-4a2d-aa20-49b0f1661914

minhthu-bic commented:
This Emoji picker looks like a Reaction modal. I've created a Emoji keyboard looks like video below. If you want, I will create a PR.


https://github.com/mattermost/mattermost-mobile/assets/97847068/aa3bb9f3-8006-4d46-8831-4705ec4894bd

enahum commented:
@abhijit-singh I actually like what @minhthu-bic has done with the keyboard area, what do you think? The only change I would make is to move the custom emojis to the end of the list.

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!
